### PR TITLE
OCPBUGS-43892: manual rebase on upstream release-0.11

### PR DESCRIPTION
**Pull request by mandre**
Supersedes https://github.com/openshift/cluster-api-provider-openstack/pull/326 and https://github.com/openshift/cluster-api-provider-openstack/pull/330

### CODE REVIEW

File: openshift/pkg/infraclustercontroller/openstackcluster.go [hunk 1]
```diff
@@ -355,6 +355,10 @@ func (r *OpenShiftClusterReconciler) getDefaultSubnetFromMachines(ctx context.Co
 		return nil, fmt.Errorf("listing control plane machines: %w", err)
 	}
 
+	if mapiMachines.Items == nil {
```
Comment (mdbooth) [inline openshift/pkg/infraclustercontroller/openstackcluster.go]: This should be
```go
  if len(mapiMachines.Items) == 0 {
```

Comparing a slice to nil is almost never correct.

**Comment by mandre** [comment]
Looking at [the logs from the last run](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-api-provider-openstack/331/pull-ci-openshift-cluster-api-provider-openstack-main-e2e-techpreview/1854347279620640768/artifacts/e2e-techpreview/gather-extra/artifacts/pods/), I see a couple of issues:

Errors in [capo-controller-manager](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-api-provider-openstack/331/pull-ci-openshift-cluster-api-provider-openstack-main-e2e-techpreview/1854347279620640768/artifacts/e2e-techpreview/gather-extra/artifacts/pods/openshift-cluster-api_capo-controller-manager-6dddd95ddf-hslhr_manager.log)
```
E1107 04:01:08.489937       1 kind.go:71] "if kind is a CRD, it should be installed before calling Start" err="no matches for kind \"Image\" in version \"openstack.k-orc.cloud/v1alpha1\"" logger="controller-runtime.source.EventHandler" kind="Image.openstack.k-orc.cloud"
```
Also some cert issues
```
2024/11/07 04:01:13 http: TLS handshake error from 10.129.0.2:33060: remote error: tls: bad certificate
```

Eventually, CAPO stops:
```
I1107 04:03:08.481743       1 server.go:249] "Shutting down webhook server with timeout of 1 minute" logger="controller-runtime.webhook"
I1107 04:03:08.482153       1 internal.go:544] "Stopping and waiting for HTTP servers"
I1107 04:03:08.482178       1 server.go:68] "shutting down server" name="health probe" addr="[::]:9440"
I1107 04:03:08.482205       1 server.go:254] "Shutting down metrics server with timeout of 1 minute" logger="controller-runtime.metrics"
I1107 04:03:08.482229       1 internal.go:548] "Wait completed, proceeding to shutdown the manager"
E1107 04:03:08.487353       1 main.go:285] "problem running manager" err="failed to wait for image caches to sync: timed out waiting for cache to be synced for Kind *v1alpha1.Image" logger="setup"
```

That likely explains the webhook errors we see in [the capi-controller-manager logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-api-provider-openstack/331/pull-ci-openshift-cluster-api-provider-openstack-main-e2e-techpreview/1854347279620640768/artifacts/e2e-techpreview/gather-extra/artifacts/pods/openshift-cluster-api_capi-controller-manager-86bfdc88d5-gl6tp_manager.log):

```
E1107 03:37:25.937789       1 controller.go:329] "Reconciler error" err="failed to update Machines: failed to update InfrastructureMachine openshift-cluster-api/openstack-machineset-zqtqb: failed to update openshift-cluster-api/openstack-machineset-zqtqb: failed to apply OpenStackMachine openshift-cluster-api/openstack-machineset-zqtqb: Internal error occurred: failed calling webhook \"validation.openstackmachine.infrastructure.cluster.x-k8s.io\": failed to call webhook: Post \"https://capo-webhook-service.openshift-cluster-api.svc:443/validate-infrastructure-cluster-x-k8s-io-v1beta1-openstackmachine?timeout=10s\": no endpoints available for service \"capo-webhook-service\"" controller="machineset" controllerGroup="cluster.x-k8s.io" controllerKind="MachineSet" MachineSet="openshift-cluster-api/openstack-machineset" namespace="openshift-cluster-api" name="openstack-machineset" reconcileID="b55c74c9-f2b5-4dd5-af07-939d897ee5ec"
```

**Comment by mdbooth** [comment]
We load the ORC CRDs from the `default` kustomization base in CAPO here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7f7f0f0fd51307bd9e87925212d6c672f625dfbd/config/default/kustomization.yaml#L16-L17

However, I vaguely recall we don't use the `default` kustomization base downstream because it pulls in too many things we don't want which we then have to take out again. i.e. It's less code to duplicate it. If that's correct, we probably just need to do the same for ORC in our downstream kustomization base.

**Comment by EmilienM** [comment]
discussed with Matt offline, we decided to ship it.

/lgtm

**Pull request merged**